sympy/polys/matrices/normalforms.py

```diff
@@ -30,12 +37,35 @@ def smith_normal_form(m):
     ...                   [ZZ(2), ZZ(16), ZZ(14)]], (3, 3), ZZ)
     >>> print(smith_normal_form(m).to_Matrix())
     Matrix([[1, 0, 0], [0, 10, 0], [0, 0, -30]])
-
     '''
-    invs = invariant_factors(m)
-    smf = DomainMatrix.diag(invs, m.domain, m.shape)
-    return smf
-
+    '''
+    Return the Smith Normal Form of a matrix `m` along with the unimodular matrices S and T.
+    '''
+    from sympy import Matrix
```

Comment:
I haven't looked through this in detail but the algorithm looks good from a quick glance.

This should definitely be implemented using DomainMatrix rather than Matrix though.